Epoll: Avoid redundant EPOLL_CTL_MOD calls

njhill opened the pull request:
Motivation

Currently an `epoll_ctl` syscall is made every time there is a change to the event interest flags (`EPOLLIN`, `EPOLLOUT`, etc) of a channel. These are only done in the event loop so can be aggregated into 0 or 1 such calls per channel prior to the next call to `epoll_wait`.

Modifications

I think further streamlining/simplification is possible but for now I've tried to minimize structural changes and added the aggregation beneath the existing flag manipulation logic.

A new `AbstractChannel#activeFlags` field records the flags last set on the epoll fd for that channel. Calls to `setFlag`/`clearFlag` update the flags field as before but instead of calling `epoll_ctl` immediately, just set or clear a bit for the channel in a new bitset in the associated `EpollEventLoop` to reflect whether there's any change to the last set value.

Prior to calling `epoll_wait` the event loop makes the appropriate `epoll_ctl(EPOLL_CTL_MOD)` call once for each channel who's bit is set.

Result

Fewer syscalls, particularly in some auto-read=false cases. Simplified error handling from centralization of these calls.

normanmaurer commented:
@netty-bot test this please

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollChannel.java:
```diff
@@ -243,31 +250,25 @@ private static boolean isAllowHalfClosure(ChannelConfig config) {
+            return;
         }
+        // schedule a task to clear the EPOLLIN as it is not safe to modify it directly
+        loop.execute(new Runnable() {
```
nit: couldn't we create this Runnable only once and re-use ?

njhill replied:
I don't think it's directly related to the change (was like this before), and it's now hit only when `setAutoRead(false)` is called from outside the EL, but sure :)

johnou replied:
wouldn't the old implementation make more sense since it's such an edge case? if it occurs often, sure makes sense, otherwise if it happens seldom the application will pay the price in memory for every channel.

normanmaurer replied:
it's only one field so I think it is ok to cache and so reduce the GC pressure

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:
```diff
@@ -59,6 +60,8 @@
     private final FileDescriptor eventFd;
     private final FileDescriptor timerFd;
     private final IntObjectMap<AbstractEpollChannel> channels = new IntObjectHashMap<AbstractEpollChannel>(4096);
+    private final BitSet pendingFlagChannels = new BitSet(); //TODO maybe just regular set of AbstractEpollChannels?
```
when we deregister the Channel from the EventLoop we also need to ensure we remove the BitSet entry.

njhill commented:
@normanmaurer have pushed a commit addressing your comments. I made the Runnable lazily constructed, not sure if that's what you had in mind or just a `final`. And still uncertain about the change since it's an extra object per channel and I guess repeated calls to `setAutoRead(false)` for a given channel aren't common?

normanmaurer commented:
@netty-bot test this please

normanmaurer commented on transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:
```diff
@@ -203,6 +207,28 @@ void add(AbstractEpollChannel ch) throws IOException {
+                        ch.modifyEvents();
+                    } catch (IOException e) {
+                        ch.pipeline().fireExceptionCaught(e);
+                        ch.close(ch.voidPromise());
```
voidPromise() is not allowed for `ch.close(...)`. Just call `ch.close()`.

njhill replied:
@normanmaurer I was just keeping this the same as the equivalent removed logic in these two places: https://github.com/netty/netty/pull/9397/files#diff-a1f23435bb30b3ac00bf72d14e710928L455 and https://github.com/netty/netty/pull/9397/files#diff-a1f23435bb30b3ac00bf72d14e710928L540. Could you confirm that those are/were incorrect?

normanmaurer replied:
using `unsafe.close(voidPromise)` is ok... while `channel.close(voidPromise)` is not.

njhill replied:
Done, but then I just noticed that `ch.unsafe().close(ch.unsafe().voidPromise())` is done elsewhere in the same class: https://github.com/netty/netty/blob/517a93d87d5c65e3c4bedf4a77183583991f32e1/transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java#L382

want me to do that instead?

normanmaurer replied:
The question is should the close traverse the pipeline or not... ?

njhill replied:
Ah right, then I guess may as well leave as-is if you agree.

normanmaurer commented:
@netty-bot test this please

normanmaurer commented:
@netty-bot test this please
